geopandas/geodataframe.py

```diff
@@ -37,6 +37,16 @@ def _geodataframe_constructor_with_fallback(*args, **kwargs):
     return df
 
 
+def _geodataframe_constructor_sliced(data=None, index=None, crs=None, **kwargs):
+    srs = pd.Series(data, index, **kwargs)
+    # Row slices of homogenous dtypes e.g df[["geometry", "geometry2"]].loc[0]
+    # pass GeometryArrays, these should stay as pd.Series
+    # https://github.com/geopandas/geopandas/issues/2282
+    if is_geometry_type(srs) and not isinstance(data, GeometryArray):
```

Comment:
In principle this is a HUGE assertion to make that this is the only place we see geometryarrays. In practice it seems most calls are either SingleBlockManagers or ndarrays, but that could be different on different pandas versions as well.

Reply:
So doing `srs.dtype == "geometry"` as a more general check would cause a row of a GeoDataFrame with only geometry dtyped columns to still be returned as a GeoSeries, while the current check gives a Series in that case?

Reply:
This is probably relying to much on the pandas internals ... I actually (for a completely unrelated reasonn, just coincidentally) just made a PR in pandas that would make that something like `df.(i)loc[0]` for a single row would _also_ pass a SingleBlockManager instead of an ndarray/GeometryArray (https://github.com/pandas-dev/pandas/pull/47077)

So that would already "break" this workaround ..

Reply:
So pandas, when constructing the array for a row, actually checks the dtypes, and if those are uniform for all columns, it will create an array of that dtype. Beause of this, we actually see a GeometryArray, and not an object dtype ndarray as I thought it would give when making the suggestion for the above solution at https://github.com/geopandas/geopandas/issues/2282#issuecomment-1004114394

So that makes that my suggestion actually doesn't work for GeoDataFrames with _only_ geometry dtype columns (actually it seems that I also already discovered that before in https://github.com/geopandas/geopandas/issues/2282#issuecomment-1025582288 .. ;))

OK, so the question is then: what do we do with selecting a single row from a geometry-only GeoDataFrames? Are we fine with making that return a GeoSeries (with the possible breaking behaviour as reported in https://github.com/geopandas/geopandas/issues/2282)? Or do we have some other (more robust) way to find out that we are selecting a row? (overriding iloc, ...)

Reply:
> So doing `srs.dtype == "geometry"` as a more general check (so only the `is_geometry_type`, without excluding GeometryArray) would cause a row of a GeoDataFrame with only geometry dtyped columns to still be returned as a GeoSeries, while the current check gives a Series in that case?

Will check again, I think this is actually for squeeze of a single geometry column geodataframe. Without this it would get demoted to a Series.